### DISCUSSION TITLE
Using ADC on ESP8266-1 to get 'system voltage'

### DISCUSSION THREAD

**Discussion by Fisherman48** [micropython/micropython#10761] [ESP8266]
Hi
Moving from lua to micropython I load fw 19.1 to a handfull of old esp8266-1 1Mb and trying out my new environment using Thonny.

I want to monitor the 'system voltage' using diff power sources.
When looking at machine.ADC( ) , it accepts 0 and 1 for parameter and I get  values I cant interprete.
ADC(1) looks like measuring on the absent /open ADC pin (all 16 bit ones) on the 8266-1 device - but the values from ADC(0) are no sense for me.

``` py 
# dif power sources
from machine import ADC
import time
def adc(m):
    for i in range(5):
        adc = ADC(m)            # create ADC object on ADC pin
        print(adc.read_u16(), adc.read()) # the 2 available methods
        time.sleep(1)
adc(0)
adc(1)

```

at 3.1 V - gives 3.1V/40=0.0775
2367 38
2367 40
2431 40
2431 41
2431 41
4194176 65535
4194176 65535
4194176 65535
4194176 65535
4194176 65535
at 3.7 V - 3.7V/120=0.031
7423 120
6463 104
7359 120
7167 116
7167 116
4194176 65535
4194176 65535
4194176 65535
4194176 65535
4194176 65535
>
1) Any hint to understand this?
2) And more abaout python: Using help( machine) and help(machine.ADC) you see ADC is a 'class' and .ADC.read it's a function - how can I go deeper - reading the class or function argument??

Br /Stefan in the Archipelago of Stockholm

**Comment by robert-hh** [comment]
AFAIK, the ESP8266 has just one 10 bit ADC with external input, with a voltage range of 0..1V. ADC(1) is for the internal Vdd voltage. It seems that one is scaled up internally to a internal 16 bit range. The ratio you get between read.u16() and read() is always 64 or 2**6.
The value of 65535 only means that the ADC((1) overflows. for ADC(0) you have to explain, what you connected to the ADC input.

**Reply by Fisherman48** [reply]
As I said there is no ADC pin on esp8266-1 module. So my intesion was to use ADC(0) to monitor the power supply (different batteries)

From nodeMCU lua about ADC:
"On the ESP8266 there is only a single-channel, which is multiplexed with the battery voltage. Depending on the setting in the "esp init data" (byte 107) one can either use the ADC to read an external voltage, or to read the system voltage (vdd33), but not both."

Where can I find the HW drawing and the documentation of class ADC ?

/stefan

**Reply by robert-hh** [reply]
The schematics of may be available at the vendor's site. But the system voltage vdd33 is a chip-internal link. The schamtics of a board will not help. The documentation of the ADC class for ESP8266 is here: https://docs.micropython.org/en/latest/esp8266/tutorial/adc.html, the C-code implementation is here: https://github.com/micropython/micropython/blob/master/ports/esp8266/machine_adc.c
I just tried it myself with an ESP8266. It seems that the ADC(1) reading is not configured. According to the reference manual, that must be done in the esp_init_data.bin, which is to be placed in the high area of the flash.

**Reply by Fisherman48** [reply]
Thanks for the C-code - great tool instead of guessing about parameters- numbers , type and repr.
How do you know/ where can I find about details like the vdd33 link ???
This link I guess, is one bit of a byte/word in the firmware - directing an analog switch between the ADC pin and a voltage divider to vdd.
Where can I dig deeper in to this?

(I'm more HW , assembler and Parallax Propeller oriented then Rasbery Py C++ and dev system person)
/Stefan

**Reply by robert-hh** [reply]
I started searching, where the setting could be changed, but did not get far. There is a esp_init_data.c file in the esp8266 file set, which is compiled into the firmware, and there is a esp_init_data.bin file coming from espressif, which is stored in flash. One of both has to be changed (or both). But how, is to be found.

**Reply by Fisherman48** [reply]
Hi - on the same road ...
esp_init_data_default.bin seems to be init/ default values and byte[107] = 255 for vdd33 or 18-36 (actual Vdd on chip pin 6 / ADC ext measurement.

"If you need to customize init data then first download the [Espressif SDK patch 1.5.4.1](http://bbs.espressif.com/download/file.php?id=1572) and extract esp_init_data_default.bin. Then flash that file just like you'd flash the firmware. The correct address for the init data depends on the capacity of the flash chip."

0xfc000 for 1 MB, modules like ESP8285, PSF-A85, some ESP-01, -03 etc.
esp_init_data_default.bin is 1024 bytes and start adress at  0xFC000 + 107 decimalt = 0xfc06b

Reading low adresses causes reboot?? The original 'esp_init_data_default.bin' is loaded at fix adress for reboot and that area is protected?
Some relocation / linker  is placing a working copy together with the rest- how find the newstart adress 

 Br /Stefan

**Reply by robert-hh** [reply]
Hi Stefan. Looks like you made a big progress. I did not look further into that topic, being busy with something else. But it looks like you will make it. The data sheet resp discussions in the espressif forum seemed to be not overly clear about what has to be written to address 107.  Yes, esp_init_data.bin has to be flashed at the top of the flash memory using esptool.py.
Best Regards, Robert

**Reply by Fisherman48** [reply]
I have 2 old? versions of the file:  esp_init_data_default.bin and esp_init_data_default_v08.bin and they differ i these positions:

1      0 8
34 82 78
35 78 74
36 74 70
37 68 64
38 64 60
114 0 1

Any ideas?
BR /Stefan

**Reply by robert-hh** [reply]
Sorry, no.

**Reply by Fisherman48** [reply]
Found the c-code for the lua adc module - I'm not fluent in C but I can see they check if the 107 byte is ok depending of vdd33 or ADC pin otherwise change the byte and reboot.
 
https://github.com/nodemcu/nodemcu-firmware/blob/release/app/modules/adc.c

Br Stefan

**Reply by robert-hh** [reply]
You can just try and see. Using esptool.py you can as well read the esp_init_default_data.bin from flash.

**Reply by Fisherman48** [reply]
Hi Robert - to close  this:
Tryed diff values on byte 107 and flashed the esp_init_default_data.bin - the byte was still intact/ same after reboot but I didnt see any change in reading voltage. So I give up digging further 'down the rabbit hole' ... 
Thanks for help /BR Stefan
